# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1725825209 -7200
# Parent  7953e450879cae8ae4f744c1da2ca7eddfe319ab
Bug NNN17 - Fix empty search terms. Allow search for |Subject is <empty>| or |To is <empty>|.

Note that "Subject" and addressing fields, like "To", are processed differently.

diff --git a/mail/base/content/SearchDialog.js b/mail/base/content/SearchDialog.js
--- a/mail/base/content/SearchDialog.js
+++ b/mail/base/content/SearchDialog.js
@@ -399,14 +399,16 @@ function getSearchTerms() {
     // A header search of "" is illegal for IMAP and will cause us to
     //  explode.  You don't want that and I don't want that.  So let's check
     //  if the bloody term is a subject search on a blank string, and if it
     //  is, let's secretly not add the term.  Everyone wins!
-    if (
-      realTerm.attrib != Ci.nsMsgSearchAttrib.Subject ||
-      realTerm.value.str != ""
-    ) {
-      searchTerms.push(realTerm);
-    }
+    // This appears incorrect since IMAP skips empty search terms anyway:
+    // https://searchfox.org/comm-central/rev/06598f8b076166b3c66bab6f71b2f6c50bcee004/mailnews/base/search/src/nsMsgImapSearch.cpp#112
+    // if (
+    //   realTerm.attrib != Ci.nsMsgSearchAttrib.Subject ||
+    //   realTerm.value.str != ""
+    // ) {
+    searchTerms.push(realTerm);
+    // }
   }
 
   return searchTerms;
 }
diff --git a/mailnews/search/src/nsMsgSearchTerm.cpp b/mailnews/search/src/nsMsgSearchTerm.cpp
--- a/mailnews/search/src/nsMsgSearchTerm.cpp
+++ b/mailnews/search/src/nsMsgSearchTerm.cpp
@@ -1089,8 +1089,22 @@ NS_IMETHODIMP nsMsgSearchTerm::MatchRfc8
 
   *pResult = false;
   bool result;
 
+  // Hand-roll comparison with empty strings.
+  // This is easier than trying to fiddle it into the processing below
+  // which splits the string into display names and e-mail addresses and
+  // compares both.
+  if (m_value.utf16String.IsEmpty()) {
+    if (m_operator == nsMsgSearchOp::Is) {
+      *pResult = string.IsEmpty();
+      return NS_OK;
+    } else if (m_operator == nsMsgSearchOp::Isnt) {
+      *pResult = !string.IsEmpty();
+      return NS_OK;
+    }
+  }
+
   // Change the sense of the loop so we don't bail out prematurely
   // on negative terms. i.e. opDoesntContain must look at all recipients
   bool boolContinueLoop;
   GetMatchAllBeforeDeciding(&boolContinueLoop);
